client/express-checkout/index.js

```diff
@@ -177,50 +188,104 @@ jQuery( ( $ ) => {
 		 * @param {Object} options ECE options.
 		 */
 		startExpressCheckoutElement: ( options ) => {
+			const getShippingRates = () => {
+				if ( ! options.requestShipping ) {
+					return [];
+				}
+
+				if ( getExpressCheckoutData( 'is_product_page' ) ) {
+					// Despite the name of the property, this seems to be just a single option that's not in an array.
+					const {
+						shippingOptions: shippingOption,
+					} = getExpressCheckoutData( 'product' );
+
+					return [
+						{
+							id: shippingOption.id,
+							amount: shippingOption.amount,
+							displayName: shippingOption.label,
+						},
+					];
+				}
+
+				return options.displayItems
+					.filter( ( i ) => i.label === 'Shipping' )
+					.map( ( i ) => ( {
+						id: `rate-${ i.label }`,
+						amount: i.amount,
+						displayName: i.label,
+					} ) );
+			};
+
+			const shippingRates = getShippingRates();
+
+			// This is a bit of a hack, but we need some way to get the shipping information before rendering the button, and
+			// since we don't have any address information at this point it seems best to rely on what came with the cart response.
+			// Relying on what's provided in the cart response seems safest since it should always include a valid shipping
+			// rate if one is required and available.
+			// If no shipping rate is found we can't render the button so we just exit.
+			if ( options.requestShipping && ! shippingRates ) {
+				return;
+			}
+
 			const elements = api.getStripe().elements( {
 				mode: options?.mode ?? 'payment',
 				amount: options?.total,
 				currency: options?.currency,
+				paymentMethodCreation: 'manual',
 			} );
 
 			const eceButton = wcpayECE.createButton( elements, {
 				buttonType: {
-					googlePay: wcpayExpressCheckoutParams.button.type,
-					applePay: wcpayExpressCheckoutParams.button.type,
+					googlePay: getExpressCheckoutData( 'button' ).type,
+					applePay: getExpressCheckoutData( 'button' ).type,
 				},
 			} );
 
 			wcpayECE.showButton( eceButton );
 
-			wcpayECE.attachButtonEventListeners( eceButton );
-
 			eceButton.on( 'click', function ( event ) {
+				// TODO: handle cases where we need login confirmation.
+
+				if ( getExpressCheckoutData( 'is_product_page' ) ) {
+					wcpayECE.addToCart();
```

Comment:
Should we `await` for the product to be added to the cart before proceeding? Alternatively, we may create a promise that resolves later.

Reply:
Se can't due to the 1 second limitation we have to resolve before we run out of time. https://docs.stripe.com/js/elements_object/express_checkout_element_click_event#express_checkout_element_on_click-handler-resolve

Reply:
Issue created here. I'll work on that later. https://github.com/Automattic/woocommerce-payments/issues/8967

Reply:
Ran some tests in https://github.com/Automattic/woocommerce-payments/issues/8967#issuecomment-2174696547 and can confirm it's failing as expected.   

Since we can't call the `resolve` function after 1 second, if adding product(s) to the cart takes too long then it should fail gracefully at least.

In the scenarios I've tested, no order is created and the user is not charged. While this is not ideal, it is an acceptable outcome considering the severity of the error.

One approach we could take is to set a flag when the request for `/?wc-ajax=wcpay_add_to_cart` is made. Then, if the shipping address or shipping options are changed, we could reject the event, as it wouldn't be necessary to continue doing that if the product hasn't been added to the cart. I'll attempt to implement this while working on issue #8871.